File: 20/python/main.py
import itertools
import math
from functools import reduce


def transform(lit_pixels_in_frame, default_pixel_outside_frame, frame, alg):
    (top_row, bottom_row), (left_col, right_col) = frame

    def window_coordinates(r, c):
        deltas = itertools.product((-1, 0, 1), repeat=2)
        return ((r + dr, c + dc) for dr, dc in deltas)

    def to_decimal(bits):
        return reduce(lambda acc, b: acc * 2 + int(b), bits)

    def transformed_pixel(r, c):
        def in_frame(r, c):
            return r >= top_row and r <= bottom_row and c >= left_col and c <= right_col

        def get_pixel(r, c):
            return (
                (r, c) in lit_pixels_in_frame
                if in_frame(r, c)
                else default_pixel_outside_frame
            )

        window = window_coordinates(r, c)
        bitstring = [get_pixel(r, c) for r, c in window]
        idx = to_decimal(bitstring)
        return alg[idx]

    result = set()
    for row in range(top_row - 2, bottom_row + 2 + 1):
        for col in range(left_col - 2, right_col + 2 + 1):
            if transformed_pixel(row, col):
                result.add((row, col))

    new_default_pixel = transformed_pixel(math.inf, math.inf)

    return result, new_default_pixel


def enhance(lit_pixels, alg, num_rows, num_cols, num_iter):
    default_outside_frame = False
    for i in range(num_iter):
        offset = 2 * i
        frame = (-offset, num_rows + offset), (-offset, num_cols + offset)
        lit_pixels, default_outside_frame = transform(
            lit_pixels, default_outside_frame, frame, alg
        )
    return len(lit_pixels)


def part1(lit_pixels, alg, num_rows, num_cols):
    return enhance(lit_pixels, alg, num_rows, num_cols, 2)


def part2(lit_pixels, alg, num_rows, num_cols):
    return enhance(lit_pixels, alg, num_rows, num_cols, 50)


with open("../input.txt") as f:
    s = f.read().strip()
    alg, img = s.split("\n\n")
    alg = "".join(alg.split())
    alg = [c == "#" for c in alg]
    img = img.split()
    num_rows, num_cols = len(img), len(img[0])
    lit_pixels = {
        (r, c) for r in range(len(img)) for c in range(len(img[r])) if img[r][c] == "#"
    }
    print(part1(lit_pixels, alg, num_rows, num_cols))
    print(part2(lit_pixels, alg, num_rows, num_cols))

File: 20/rust/src/main.rs
use std::collections::HashSet;

use bitvec::prelude::*;
use bitvec::{array::BitArray, bitarr};
use itertools::Itertools;

type Alg = BitArray<LocalBits, [usize; 8]>;

#[derive(Debug)]
struct Frame {
    top_row: i32,
    bottom_row: i32,
    left_col: i32,
    right_col: i32,
}

impl Frame {
    fn expand(&self) -> Frame {
        Frame {
            top_row: self.top_row - 2,
            bottom_row: self.bottom_row + 2,
            left_col: self.left_col - 2,
            right_col: self.right_col + 2,
        }
    }

    fn pixels(
        &self,
    ) -> itertools::Product<std::ops::RangeInclusive<i32>, std::ops::RangeInclusive<i32>> {
        (self.top_row..=self.bottom_row).cartesian_product(self.left_col..=self.right_col)
    }
}

#[derive(Debug)]
struct LazyImage {
    lit_pixels_in_frame: HashSet<(i32, i32)>,
    frame: Frame,
    lit_outside_frame: bool,
}

impl LazyImage {
    fn enhance(&mut self, alg: Alg) {
        let expanded_frame = self.frame.expand();
        let new_lit_pixels_in_frame: HashSet<(i32, i32)> = expanded_frame
            .pixels()
            .filter(|(row, col)| self.enhance_pixel(alg, *row, *col))
            .collect();
        let outside_window_value = if self.lit_outside_frame { 511 } else { 0 };
        let lit_outside_frame = *alg.get(outside_window_value).unwrap();
        self.frame = expanded_frame;
        self.lit_pixels_in_frame = new_lit_pixels_in_frame;
        self.lit_outside_frame = lit_outside_frame;
    }

    fn get_pixel(&self, row: i32, col: i32) -> bool {
        if row < self.frame.top_row
            || row > self.frame.bottom_row
            || col < self.frame.left_col
            || col > self.frame.right_col
        {
            self.lit_outside_frame
        } else {
            self.lit_pixels_in_frame.contains(&(row, col))
        }
    }

    fn window_values(&self, row: i32, col: i32) -> [bool; 9] {
        [
            (row - 1, col - 1),
            (row - 1, col),
            (row - 1, col + 1),
            (row, col - 1),
            (row, col),
            (row, col + 1),
            (row + 1, col - 1),
            (row + 1, col),
            (row + 1, col + 1),
        ]
        .map(|(row, col)| self.get_pixel(row, col))
    }

    fn enhance_pixel(&self, alg: Alg, row: i32, col: i32) -> bool {
        fn to_decimal(bitarray: [bool; 9]) -> usize {
            bitarray
                .iter()
                .fold(0, |acc, elem| acc * 2 + (*elem as usize))
        }
        let window = self.window_values(row, col);
        let idx = to_decimal(window);
        *alg.get(idx).unwrap()
    }
}

fn main() -> anyhow::Result<()> {
    let (alg, input) = (include_str!("../../input.txt"))
        .split_once("\n\n")
        .expect("invalid input");

    let alg = parse_alg(alg);
    let mut img = parse_img(input);
    println!("part 1: {}", part1(&mut img, alg));
    println!("part 2: {}", part2(&mut img, alg));
    Ok(())
}

fn part1(img: &mut LazyImage, alg: Alg) -> usize {
    img.enhance(alg);
    img.enhance(alg);
    img.lit_pixels_in_frame.len()
}

fn part2(img: &mut LazyImage, alg: Alg) -> usize {
    for _ in 0..48 {
        img.enhance(alg);
    }
    img.lit_pixels_in_frame.len()
}

fn parse_alg(s: &str) -> BitArray<LocalBits, [usize; 8]> {
    let mut result = bitarr![0; 512];
    s.lines()
        .flat_map(|line| line.bytes())
        .enumerate()
        .for_each(|(i, val)| {
            if val == b'#' {
                result.set(i, true);
            }
        });
    result
}

fn parse_img(s: &str) -> LazyImage {
    let mut lit_pixels = HashSet::new();
    let lines = s.lines().map(|line| line.as_bytes()).collect_vec();
    for (r, row) in lines.iter().enumerate() {
        for c in 0..lines[r].len() {
            if row[c] == b'#' {
                lit_pixels.insert((r as i32, c as i32));
            }
        }
    }
    LazyImage {
        lit_pixels_in_frame: lit_pixels,
        frame: Frame {
            top_row: 0,
            bottom_row: lines.len() as i32 - 1,
            left_col: 0,
            right_col: lines[0].len() as i32 - 1,
        },
        lit_outside_frame: false,
    }
}

fn _parse_img(s: &str) -> Vec<BitVec> {
    s.lines()
        .map(|line| {
            let line = line.as_bytes();
            let mut result = bitvec![0; line.len()];
            for (i, c) in line.iter().enumerate() {
                if *c == b'#' {
                    result.set(i, true);
                }
            }
            result
        })
        .collect_vec()
}
